chainer/iterators/serial_iterator.py

```diff
@@ -30,15 +33,34 @@ class SerialIterator(iterator.Iterator):
         shuffle (bool): If ``True``, the order of examples is shuffled at the
             beginning of each epoch. Otherwise, examples are extracted in the
             order of indexes.
+        order_sampler (callable): A callable that generates the order
+            of the indices to sample in the next epoch when a epoch finishes.
+            This function should take two arguements: the current order
+            and the current position of the iterator.
+            This should return the next order. The size of the order
+            should remain constant.
+            This option can not be used when ``shuffle`` is ``True``.
 
     """
 
-    def __init__(self, dataset, batch_size, repeat=True, shuffle=True):
+    def __init__(self, dataset, batch_size,
+                 repeat=True, shuffle=True, order_sampler=None):
         self.dataset = dataset
         self.batch_size = batch_size
         self._repeat = repeat
         self._shuffle = shuffle
 
+        if self._shuffle and order_sampler is not None:
+            warnings.warn('`shuffle` is `True` and a custom '
```

Comment:
I think we can be more strict and simply raise an error, since we don't need this for backward compatibility?